logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java

```diff
@@ -0,0 +1,438 @@
+         * trivial dataset that does not invoke any computation whatsoever.</p>
+         * The return of a call to
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)} is always
+         * {@link Collections#emptyList()}.
+         * @param parents Parent {@link Dataset} to sum and terminate
+         * @return Dataset representing the sum of given parent {@link Dataset}
+         */
+        public static Dataset from(final Collection<Dataset> parents, final boolean debug) {
+            final int count = parents.size();
+            final Dataset result;
+            if (count > 0) {
+                if (debug) {
+                    result = new Dataset.TerminalDataset.TerminalDebugDataset(parents);
+                } else {
+                    result = new Dataset.TerminalDataset(parents);
+                }
+            } else {
+                result = TRIVIAL;
+            }
+            return result;
+        }
+
+        private TerminalDataset(final Collection<Dataset> parents) {
+            this.parents = parents;
+        }
+
+        @Override
+        public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+            final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+            final RubyHash options) {
+            parents.forEach(dataset -> dataset.compute(batch, flush, shutdown, options));
+            this.clear();
+            return EMPTY_RETURN;
+        }
+
+        @Override
+        public void clear() {
+            for (final Dataset parent : parents) {
+                parent.clear();
+            }
+        }
+
+        /**
+         * <p>{@link Dataset} that contains all {@link JrubyEventExtLibrary.RubyEvent} of all of
+         * its dependencies and is assumed to be at the end of an execution.</p>
+         * This dataset does not require an explicit call to {@link Dataset#clear()} since it will
+         * automatically {@code clear} all of its parents.
+         * Unlike {@link Dataset.TerminalDataset} this class will return all
+         * {@link JrubyEventExtLibrary.RubyEvent} that passed through it on a call to
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)}.
+         */
+        public static final class TerminalDebugDataset implements Dataset {
+
+            private final Collection<Dataset> parents;
+
+            private TerminalDebugDataset(final Collection<Dataset> parents) {
+                this.parents = parents;
+            }
+
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+                final RubyHash options) {
+                final Collection<JrubyEventExtLibrary.RubyEvent> res = new ArrayList<>(10);
+                parents.forEach(
+                    dataset -> res.addAll(dataset.compute(batch, flush, shutdown, options))
```

Comment:
The code we're adding for debugging seems, at a glance, fairly complex. The difference between TerminalDebugDataset and TerminalDataset seems to be that one returns EMPTY_RETURN and the other returns a collection of all(?) RubyEvents? Why? What's the story for this feature?

Reply:
@jordansissel this is just another outgrowth of what we need for working with dev utils at the moment.
Fixing dev-utils is kinda meh ... I tried but realized I need more functionality in there as well to test the exceptions ... will see what I can do about this today.

In general though ... we should really move them into core imo, a different release cycle for your UT utils is just kinda painful :)?